### ISSUE TITLE
256/16 million color support

### ISSUE THREAD

**Issue by Qix-** [chalk/ansi-styles#11]
Yes, [it's time](https://twitter.com/sindresorhus/status/613460287270842368).

I spent the greater part of last night trying to figure out the best way to tackle compatibility with ANSI 256/16m (truecolor) support to where colors would degrade nicely. I started [writing my own](https://github.com/Qix-/ink) engine but alas I figured this may be better suited as a PR for an existing solution (i.e. Node's biggest package, Chalk). Now that `supports-color` [will soon support more colors](https://github.com/sindresorhus/supports-color/pull/25), this is a good time to look at it.

There are a few issues issues I foresee with an upgrade such as this:

- What happens to 256/16m colors on terminals that only support 16 colors (_"basic"_ support)? 16m colors on terminals that only support 256 colors?
- How do we accurately and easily wrap color operations into an already expressive API, with full backwards compatibility?

## Color Operations
The way my engine does it (and the way that seems to work fine) is to use the [`color-js`](https://www.npmjs.com/package/color-js) package to perform color operations on full color values. This includes modifying lightness, saturation, and per-channel values.

These color values can then be converted to 8bpc RGB color values, which can then be used to degrade into the lower forms.

### Tests
First of all, let's define our domain scaling function. This function takes a value, along with the upper/lower bound of that value's domain, and then applies it to a lower/upper bound of another domain.

~~~coffeescript
applyDomain = (val, lb, ub, tlb, tub)->
  Math.round ((val - lb) / (ub - lb)) * (tub - tlb) + tlb
~~~

From there, degrading colors becomes somewhat of a simple task.

For each of the three [levels](https://github.com/sindresorhus/supports-color/pull/25#discussion_r33014946), I'll be decoding an image using [png-js](https://github.com/devongovett/png.js) and running the pixel data through each of the functions.

The original image:
![applepride](https://cloud.githubusercontent.com/assets/885648/8319567/806bb6c6-19cd-11e5-9a6a-0ab623c52716.png)

#### 16m (8bpc)
This one is simple. Provided a 0..255 domain of RGB values, we can then build up a pretty lossless result in the terminal.

> color-js provides RGB values in floating point numbers in the domain 0.0..1.0. As you might expect, this is easily converted to the 0..255 domain using `applyDomain n, 0, 1, 0, 255`.

![screen shot 2015-06-23 at 5 10 51 pm](https://cloud.githubusercontent.com/assets/885648/8319329/d5bffda6-19ca-11e5-8ca8-1e2bb294fabb.png)

#### 16m (8bpc) to ANSI 256
Thanks to [this formula](https://github.com/jbnicolai/ansi-256-colors/blob/ef02d765885057e592cd1e9fa400003ada486f91/index.js#L30) (thanks @jbnicolai) we know the per-channel domain of ANSI-256 colors is 0..5 with a translation of +16.

~~~coffeescript
toAnsi256 = (r8, g8, b8)->
  r = applyDomain r8, 0, 255, 0, 5 # red, from 0..255 to 0..5
  g = applyDomain g8, 0, 255, 0, 5 # green, from 0..255 to 0..5
  b = applyDomain b8, 0, 255, 0, 5 # blue, from 0.255 to 0..5
  return (36 * r) + (6 * g) + b + 16
~~~

As you can see, it's quite accurate.

![screen shot 2015-06-23 at 5 05 56 pm](https://cloud.githubusercontent.com/assets/885648/8319331/df6a7cdc-19ca-11e5-9bd2-106d66240ec8.png)

#### 16m (8bpc) to 16 colors
This one was a little trickier, but really achieves the same effect.

~~~coffeescript
toAnsi16 = (r8, g8, b8)->
  r = applyDomain r8, 0, 255, 0, 1 # red, from 0..255 to 0..1 (rounded)
  g = applyDomain g8, 0, 255, 0, 1 # green, from 0..255 to 0..1 (rounded)
  b = applyDomain b8, 0, 255, 0, 1 # blue, from 0..255 to 0..1 (rounded)
  return (b << 2) | (g << 1) | r
~~~

this yields predictable results.

~~~
coffee> toAnsi16 = require './ansi16'
[Function]
coffee> toAnsi16 255, 255, 255
7
coffee> toAnsi16 255, 0, 0
1
coffee> toAnsi16 255, 0, 126
1
coffee> toAnsi16 255, 0, 127
1
coffee> toAnsi16 255, 0, 128
5
coffee> toAnsi16 255, 127, 128
5
coffee> toAnsi16 30, 185, 128
6
~~~

As you can see, it's also pretty accurate (~~minus orange, which is a tricky color to get right with the standard 16 color palette~~).

![screen shot 2015-06-23 at 5 03 16 pm](https://cloud.githubusercontent.com/assets/885648/8319267/fee3305a-19c9-11e5-826f-e1d4def0a5d7.png)

Though by using color-js and the lightness value bound to the domain 0..1 (rounded), then we can select the brightness scheme using the extended `100m`-`107m` bright background codes to get a better result.

![screen shot 2015-06-23 at 5 43 10 pm](https://cloud.githubusercontent.com/assets/885648/8319723/561efd68-19cf-11e5-9f5b-1a0a62e98bac.png)

Much better - and with only the basic functionality (and yes, [even Windows supports those](https://github.com/joyent/node/blob/master/deps/uv/src/win/tty.c#L1233-L1238)).

### Internal structure
With the domain functions being pretty accurate, we can then start making assertions about color values. This means we can store color-js `Color` objects as the main source for color, perform operations on them, and then degrade them down to corresponding command line colors.

For instance, here is my prototype library doing this with a nice shade of Purple. There isn't a ton of logic about the 16 color palette, but I'm sure something could be thrown in there for lightness governing the bold (`1m`) code.

![screen shot 2015-06-23 at 5 18 26 pm](https://cloud.githubusercontent.com/assets/885648/8319424/f48fa1ea-19cb-11e5-9766-b7d1d2e31926.png)

---

As you can see in the above prototype, we now have a *lot* of control over the kinds of colors we can output to the console. We can do advanced things like shift by hue, lighten/darken by lightness/value, and even start theming console colors with CSS (if we so choose) - all while being completely degradable by non-supporting terminals.

Existing programs can upgrade their color suites to be more colorful, and if we so choose [supports-color](https://github.com/sindresorhus/supports-color) could even [include or modify an environment variable](https://github.com/Qix-/supports-color/blob/256-color/index.js#L55) to allow users to set this in their RC files to enable system-wide 256 or 16 million color terminal applications.

Thoughts? I'm fully prepared to make a complete PR for this.

**Comment by jbnicolai** [comment]
Just a quick note (I'm traveling atm) that I did something similair a while back. You can probably find the discussion under chalk's issues and see my implementation here: https://github.com/jbnicolai/ansi-256-colors

**Comment by Qix-** [comment]
@jbnicolai I stole the RGB -> ANSI 256 formula from there :wink: This proposal does something a little more than just that, though. If you wanted to, we could beef up the ansi-256-colors repository to do the color-to-ANSI-code conversions and then let this library use that? Though *ansi-256-colors* is a bit of a misnomer since we'd also be supporting truecolor as well.

**Comment by jbnicolai** [comment]
Nah, I'm perfectly happy retiring that repository and beefing up the chalk/ansi-styles/supports-color stack, although I'm starting to think we should maybe move this set of repositories to their own organisation.

How would you feel about that, @sindresorhus?

**Comment by Qix-** [comment]
http://github.com/chalk seems inactive. Github is pretty good about releasing inactive accounts. http://github.com/chalkjs is also unclaimed.

**Comment by jbnicolai** [comment]
@Qix- yeah, completely agree those should probably come with as well. It's @sindresorhus' call though :)

**Comment by Qix-** [comment]
:+1:

**Comment by Qix-** [comment]
I'm looking at the existing code for this repo. It's pretty basic. It'd be pretty much a complete re-write.

**Comment by Qix-** [comment]
I've started a bit of a re-write as it is. Any thoughts on this @sindresorhus?

**Comment by sindresorhus** [comment]
I used up all my time today moving everything over to the `chalk` org. Will get back to this in a couple of days when I have more time. [Tweeted](https://twitter.com/sindresorhus/status/615969432163217408) about getting feedback on this, so we can hopefully get some more eyes on it.

**Comment by Qix-** [comment]
Sounds good :+1:

**Comment by sunesimonsen** [comment]
My package https://github.com/sunesimonsen/magicpen supports 256 with backwards compatibility with 16 color terminals. Maybe that could be a source of inspiration, the ansi serializer is here: https://github.com/sunesimonsen/magicpen/blob/master/lib/AnsiSerializer.js

**Comment by Qix-** [comment]
@sunesimonsen and what of 16m colors?

**Comment by sunesimonsen** [comment]
@Qix- we don't support that yet, but I'm curious to see what you come up with.

**Comment by Qix-** [comment]
@sunesimonsen as in the original post here, we'd be using full RGB values and degrading into their lower forms (truecolor -> 256 -> 16 colors).

**Comment by Qix-** [comment]
x-post: chalk/chalk#73

**Comment by Qix-** [comment]
Also, another question is: where should everything go?

We have a few moving parts here:

- [Level](https://github.com/chalk/supports-color/pull/25/files#diff-0730bb7c2e8f9ea2438b52e419dd86c9R16) detection
- Color manipulation functionality (hue shifting, lightness adjustments, RGB <-> HSV/HSL conversions, etc.)
- Degradation from truecolor to 256/basic values
- The fact we want to be able to map these to basic colors (i.e. `rgb(255, 0, 0)` should map to `31` instead of `38;2;255;0;0`)
- Listing above combinations (and applying labels to them)
- Styles (bold, italic, underline, inverse, etc.)
- The API for stringing them together

Some of the above are obvious (i.e. *stringing them together* would occur in chalk proper), but the others?

As well, this has the potential to introduce a **major** performance hit if not done correctly (color modifications aren't nearly as quick as lookups). Memoization will be important here.

I propose the following:

- **Level Detection**: [supports-color](/chalk/supports-color)
- **Color Manipulation**: [ansi-styles](/chalk/ansi-styles) / new module (see note below)
- **Degradation**: [ansi-styles](/chalk/ansi-styles) / new module (see note below)
- **Color Mapping**: [ansi-styles](/chalk/ansi-styles)
- **Map Labels**: [ansi-styles](/chalk/ansi-styles)
- **Styles**: [ansi-styles](/chalk/ansi-styles)
- **API**: [chalk](/chalk/chalk)

---

Color manipulation (lightness, hue, etc.) was [originally](https://github.com/chalk/ansi-styles/issues/11) going to happen with [color-js](https://github.com/brehaut/color-js). However, upon actually looking at the code, it doesn't look very performant / memory conscious. The code is under the BSD clause 2, which means we can use the algorithms with attribution and build our own structure around it.

But that's all up to @sindresorhus.

**Comment by Qix-** [comment]
Oh I'm not done.

Another consideration: if we go from high->low resolution on colors, the 16 color palette is going to be tricky.

[There are a few ways we can interpret RGB values for the 16 color palette.

![screen shot 2015-06-30 at 11 32 25 pm](https://cloud.githubusercontent.com/assets/885648/8448101/6c6aa6f6-1f80-11e5-80f5-d8806c4221c8.png)](https://en.wikipedia.org/wiki/ANSI_escape_code#Colors)

By the math, regular colors should be `0..127` and bright colors should be `128..255`, though that's not necessarily true on all VGA structures. Do we enforce this idea? People usually have them re-mapped anyway.

**Comment by sunesimonsen** [comment]
@Qix- I also using full RGB values and degrading them to the 256 and 16 color palette. I wanted 256 colors to work in terminals that supports 256 colors but can't be detected as such, so we wrap the 256 color escape sequence in a 16 color escape sequence. Old terminals will ignore the 256 color escape sequence and fallback to the 16 color escape sequence. It is a bit hacky, but it works.

**Comment by Qix-** [comment]
@sunesimonsen Ehh I feel like that's a workaround that will backfire on some terminals. Plus you're crapping up the terminal with more data than it needs - TTY operations are expensive as it is.

**Comment by sunesimonsen** [comment]
@Qix- I'm not saying you should copy that, I'm just telling you want I have done.

**Comment by jbnicolai** [comment]
@sunesimonsen that's pretty awesome actually :smile:  - although I don't suggest we copy that here

**Comment by Qix-** [comment]
The fact is the color operations will be easy, it's the API that's troublesome.

**Comment by Qix-** [comment]
I've made a PR for `color-convert` for ansi16/256 color degradation algorithms.

harthur/color-convert#18

**Comment by MoOx** [comment]
PR as been merged and released as color-convert@0.6.0.

**Comment by Qix-** [comment]
Thank you @MoOx :)
